daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Expression.scala

```diff
@@ -1146,10 +1156,8 @@ case class NamedStep(s: String, predArg: Option[PredicateExpression])
         new DownArray(nqn)
       }
     } else {
-      //
-      // Note: DFDL spec allows a[exp] if a is not an array, but it's a processing
-      // error if exp doesn't evaluate to 1.
-      // TODO: Implement this.
+      // a[exp] is only supported on arrays , because Daffodil no longer treats
+      // optional elements as arrays
       if (pred.isDefined)
```

Comment:
It looks like subset errors were intended for things we chose not to implement until a future release...since we are not planning on implementing this functionality (i.e .[1] or ..[1] or OptionalElement[1]) in the future...I'll get rid of the subset error, but keep the original error message "Indexing is only allowed on arrays. Offending path step: '%s%s'."